lib/checkunusedvar.cpp

```diff
@@ -1586,3 +1589,69 @@ bool CheckUnusedVar::isEmptyType(const Type* type)
     emptyType=false;   // unknown types are assumed to be nonempty
     return emptyType;
 }
+
+bool CheckUnusedVar::isFunctionWithoutSideEffects(const Function& func, const Token& functionUsageToken) {
+    // no body to analyze
+    if (!func.hasBody()) {
+        return false;
+    }
+
+    for (const Token* argsToken = functionUsageToken.next(); !Token::simpleMatch(argsToken, ")"); argsToken = argsToken->next()) {
+        const Variable* argVar = argsToken->variable();
+        if (argVar && argVar->isGlobal()) {
+            return false; // TODO: analyze global variable usage
+        }
+    }
+
+    bool sideEffectReturnFound = false;
+    for (Token* bodyToken = func.functionScope->bodyStart->next(); bodyToken != func.functionScope->bodyEnd; 
+        bodyToken = bodyToken->next())
+    {
+        const Variable* bodyVariable = bodyToken->variable();
+        if (bodyVariable) {
+            // check variable for side-effects
+            if (!isVariableWithoutSideEffects(*bodyVariable)) {
+                return false;
+            }
+            // check if global variable is changed
+            if (bodyVariable->isGlobal()) { 
+                return false; // TODO: analyze global variable usage
+            }
+        }
+
+        // check nested function
+        const Function* bodyFunction = bodyToken->function();
+        if (bodyFunction) {
+            if (bodyFunction->fullName() == func.fullName()) { // recursion found
```

Comment:
This will only prevent recursion when a function calls itself directly as far as I see.
I would suggest that you save nested functions in a `std::list<const Function *>` list. The list has to be passed as a parameter to `isFunctionWithoutSideEffects`.